runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -639,6 +640,164 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
       {
       switch (rm)
          {
+         case TR::java_lang_invoke_MethodHandle_asType:
+            {
+            TR::Node* mh = node->getArgument(0);
+            TR::Node* mt = node->getArgument(1);
+            bool mhConstraintGlobal, mtConstraintGlobal;
+            TR::VPConstraint* mhConstraint = getConstraint(mh, mhConstraintGlobal);
+            TR::VPConstraint* mtConstraint = getConstraint(mt, mtConstraintGlobal);
+            J9VMThread* vmThread = comp()->fej9()->vmThread();
+            TR_OpaqueClassBlock* methodHandleClass = J9JitMemory::convertClassPtrToClassOffset(J9VMJAVALANGINVOKEMETHODHANDLE(vmThread->javaVM));
+            TR_OpaqueClassBlock* methodTypeClass = J9JitMemory::convertClassPtrToClassOffset(J9VMJAVALANGINVOKEMETHODTYPE(vmThread->javaVM));
+            if (mhConstraint
+                && mtConstraint
+                && mhConstraint->getKnownObject()
+                && mhConstraint->isNonNullObject()
+                && mtConstraint->getKnownObject()
+                && mtConstraint->isNonNullObject()
+                && mhConstraint->getClass()
+                && mtConstraint->getClass()
+                && comp()->fej9()->isInstanceOf(mhConstraint->getClass(), methodHandleClass, true, true) == TR_yes
+                && comp()->fej9()->isInstanceOf(mtConstraint->getClass(), methodTypeClass, true, true) == TR_yes)
+               {
+               uintptrj_t* mhLocation = getObjectLocationFromConstraint(mhConstraint);
+               uintptrj_t* mtLocation = getObjectLocationFromConstraint(mtConstraint);
+               uint32_t mtOffset = J9VMJAVALANGINVOKEMETHODHANDLE_TYPE_OFFSET(vmThread);
+
+               TR::VMAccessCriticalSection asType(comp(),
+                                                           TR::VMAccessCriticalSection::tryToAcquireVMAccess);
+               if (!asType.hasVMAccess())
+                  break;
+               uintptrj_t mhObject = comp()->fej9()->getStaticReferenceFieldAtAddress((uintptrj_t)mhLocation);
+               uintptrj_t mtInMh = comp()->fej9()->getReferenceFieldAtAddress(mhObject + mtOffset);
+               uintptrj_t mtObject = comp()->fej9()->getStaticReferenceFieldAtAddress((uintptrj_t)mtLocation);
+
+               if (trace())
+                  traceMsg(comp(), "mhLocation %p mtLocation %p mhObject %p mtOffset %d mtInMh %p mtObject %p\n", mhLocation, mtLocation, mhObject, mtOffset, mtInMh, mtObject);
+
+               // AOT doesn't relocate/validate objects, which is why this transformation is disabled under AOT
+               if (mtInMh == mtObject)
+                  {
+                  if (!performTransformation(comp(), "%sTransforming %s on node %p to a PassThrough with child %p\n", OPT_DETAILS, signature, node, mh))
+                     break;
+
+                  anchorAllChildren(node, _curTree);
+                  node->removeAllChildren();
+                  node = TR::Node::recreateWithoutProperties(node, TR::PassThrough, 1, mh);
```

Comment:
This looks just like `transformCallNodeToPassThrough()`, which is called in the other new cases. Might it make sense to relax the assertion in that function somehow instead of duplicating its code here?

Reply:
I have an OMR change to extend the usage of that function. Please check https://github.com/eclipse/omr/pull/2679. Will make the change in openj9 shortly.